src/stability_sdk/animation.py

```diff
@@ -780,24 +787,34 @@ def _span_render_frame(self, frame_idx: int, init: Optional[np.ndarray], mask: O
         )
         return results[generation.ARTIFACT_IMAGE][0]
 
-    def _span_render(self, start: int, end: int, prev_frame: Optional[np.ndarray]) -> Generator[Tuple[int, np.ndarray], None, None]:
+    def _span_render(self, start: int, end: int, seed: int, prev_frame: Optional[np.ndarray]) -> Generator[Tuple[int, np.ndarray], None, None]:
         args = self.args
 
-        def apply_xform(frame: np.ndarray, xform: matrix.Matrix) -> Tuple[np.ndarray, np.ndarray]:
+        def apply_xform(frame: np.ndarray, xform: matrix.Matrix, frame_idx: int) -> Tuple[np.ndarray, np.ndarray]:
+            args, frame_args = self.args, self.frame_args
             if args.animation_mode == '2D':
                 xform = to_3x3(xform)
-            frames, masks = self.api.transform([frame], resample_op(args.border, xform, export_mask=True))
+                frames, masks = self.api.transform([frame], resample_op(args.border, xform, export_mask=True))
+            else:
+                fov = frame_args.fov_series[frame_idx]
+                depth_blur = int(frame_args.depth_blur_series[frame_idx])
+                depth_warp = frame_args.depth_warp_series[frame_idx]
+                projection = matrix.projection_fov(math.radians(fov), 1.0, args.near_plane, args.far_plane)                
+                wvp = matrix.multiply(projection, xform)
```

Comment:
I feel like initialisation of `wvp` and `projection` matrices should be done outside of internal `apply_xform` function, so that `apply_xform` follows `resample_op` syntax of `transform, previous_transform=None` parameters. What do you think?

Reply:
That sounded good to me so gave it a shot but it turned out worse I think. This `apply_xform` is a small utility closure within the `_span_render` method that is called to apply the accumulated forward and backward transforms. Having it in a closure avoids the code duplication between the two. I thought the proposed change would be good to eliminate needing to pass in frame index and matching resample_op as you say but the frame index is still needed to get the per-frame depth settings so it would just push the projection matrix building out of the closure to the two callers and still need the rest.